ws-nextjs-app/pages/_app.page.tsx

```diff
@@ -127,3 +136,55 @@ export default function App({ Component, pageProps }: Props) {
     </ToggleContextProvider>
   );
 }
+
+const addServiceChainAndCspHeaders = async ({
+  ctx,
+  service,
+  toggles,
+}: CspHeaderResponseProps) => {
+  ctx.res?.setHeader(
+    'req-svc-chain',
+    addPlatformToRequestChainHeader({
+      headers: ctx.req?.headers as unknown as Headers,
+    }),
+  );
+
+  const hostname = ctx.req?.headers.host || '';
+
+  const LOCALHOST_DOMAINS = ['localhost', '127.0.0.1'];
+
+  const isLocalhost = LOCALHOST_DOMAINS.includes(hostname.split(':')?.[0]);
+
+  const PRODUCTION_ONLY = !isLocalhost && process.env.NODE_ENV === 'production';
+
+  if (PRODUCTION_ONLY) {
+    await cspHeaderResponse({ ctx, service, toggles });
+  }
+};
+
+App.getInitialProps = async ({ ctx }: AppContext) => {
+  const {
+    req,
+    asPath,
+    query: { service },
+  } = ctx as NextPageContext & { query: PageDataParams };
+
+  const toggles = await getToggles(service);
+
+  await addServiceChainAndCspHeaders({ ctx, service, toggles });
+
+  const { isApp, isAmp, isLite } = getPathExtension(asPath || '');
+
+  return {
+    pageProps: {
+      // These are props passed down to ALL pages and merged with page
+      // specific props from getInitialProps / getServerSideProps
+      ...extractHeaders(req?.headers || {}),
```

Comment:
`getInitialProps` calls `extractHeaders(req?.headers || {})` but on client navigations is `ctx.req` undefined? Then `isUK/showAds/cookie banner` props might reset to defaults. Do we need a client-safe source or to keep passing these via page `getServerSideProps` to avoid state switching on client route changes?

Reply:
Its a good point to highlight though, from the [docs](https://nextjs.org/docs/pages/api-reference/functions/get-initial-props) :


> If getInitialProps is used in a custom _app.js, and the page being navigated to is using getServerSideProps, then getInitialProps will only run on the server.